flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroDeserializationSchema.java

```diff
@@ -0,0 +1,215 @@
+
+/**
+ * Deserialization schema that deserializes from Avro binary format.
+ *
+ * @param <T> type of record it produces
+ */
+public class AvroDeserializationSchema<T> implements DeserializationSchema<T> {
+
+	/**
+	 * Class to deserialize to.
+	 */
+	private Class<T> recordClazz;
+
+	private String schemaString = null;
+
+	/**
+	 * Reader that deserializes byte array into a record.
+	 */
+	private transient GenericDatumReader<T> datumReader;
+
+	/**
+	 * Input stream to read message from.
+	 */
+	private transient MutableByteArrayInputStream inputStream;
+
+	/**
+	 * Avro decoder that decodes binary data.
+	 */
+	private transient Decoder decoder;
+
+	/**
+	 * Avro schema for the reader.
+	 */
+	private transient Schema reader;
+
+	/**
+	 * Creates a Avro deserialization schema.
+	 *
+	 * @param recordClazz class to which deserialize. Should be one of:
+	 *                    {@link org.apache.avro.specific.SpecificRecord},
+	 *                    {@link org.apache.avro.generic.GenericRecord}.
+	 * @param reader      reader's Avro schema. Should be provided if recordClazz is
+	 *                    {@link GenericRecord}
+	 */
+	AvroDeserializationSchema(Class<T> recordClazz, @Nullable Schema reader) {
+		Preconditions.checkNotNull(recordClazz, "Avro record class must not be null.");
+		this.recordClazz = recordClazz;
+		this.inputStream = new MutableByteArrayInputStream();
+		this.decoder = DecoderFactory.get().binaryDecoder(inputStream, null);
+		this.reader = reader;
+		if (reader != null) {
+			this.schemaString = reader.toString();
+		}
+	}
+
+	/**
+	 * Creates {@link AvroDeserializationSchema} that produces {@link GenericRecord} using provided schema.
+	 *
+	 * @param schema schema of produced records
+	 * @return deserialized record in form of {@link GenericRecord}
+	 */
+	public static AvroDeserializationSchema<GenericRecord> forGeneric(Schema schema) {
+		return new AvroDeserializationSchema<>(GenericRecord.class, schema);
+	}
+
+	/**
+	 * Creates {@link AvroDeserializationSchema} that produces classes that were generated from avro schema.
+	 *
+	 * @param tClass class of record to be produced
+	 * @return deserialized record
+	 */
+	public static <T extends SpecificRecord> AvroDeserializationSchema<T> forSpecific(Class<T> tClass) {
+		return new AvroDeserializationSchema<>(tClass, null);
+	}
+
+	GenericDatumReader<T> getDatumReader() {
+		if (datumReader != null) {
+			return datumReader;
+		}
+
+		if (SpecificRecord.class.isAssignableFrom(recordClazz)) {
+			this.datumReader = new SpecificDatumReader<>();
+		} else if (GenericRecord.class.isAssignableFrom(recordClazz)) {
+			this.datumReader = new GenericDatumReader<>();
+		} else {
+			this.datumReader = new ReflectDatumReader<>();
+		}
+
+		return datumReader;
+	}
+
+	Schema getReaderSchema() {
+		if (reader != null) {
+			return reader;
+		}
+
+		if (SpecificRecord.class.isAssignableFrom(recordClazz)) {
+			this.reader = SpecificData.get().getSchema(recordClazz);
```

Comment:
Do we need to pass the use code class loader into the SpecificData / GenericData / etc? In order to resolve nested user-defined types?

Reply:
Didn't think about it, sorry.